osu.Game/Graphics/Containers/SectionsContainer.cs

```diff
@@ -177,36 +189,104 @@ protected override void UpdateAfterChildren()
+                return;
+            }
+
+            var sortedByVisibility = new List<T>(Children);
+            sortedByVisibility.Sort(compareVisibility);
+            SelectedSection.Value = sortedByVisibility.FirstOrDefault();
+        }
+
+        private int compareVisibility(T a, T b)
+        {
+            // note: the decision to prioritize higher sections and differentiate scroll direction
+            // comes from the fact that our scroll target is the top of the screen
+            //
+            // if this is ever not the case, this comparison should be rewritten
+            float aIndex = IndexOf(a);
+            float bIndex = IndexOf(b);
+            float aVisibleHeight = visibleHeight(a);
+            float bVisibleHeight = visibleHeight(b);
+            bool aAllVisible = Precision.AlmostBigger(aVisibleHeight / a.Height, 1.0f);
+            bool bAllVisible = Precision.AlmostBigger(bVisibleHeight / b.Height, 1.0f);
+
+            // if one is all visible while the other isn't
+            if (aAllVisible != bAllVisible)
+            {
+                // don't make the 100% visible section sort earlier in the list if it's not already a higher section
+                // this is so we don't select a small section as soon as it gets on screen (only unwanted while scrolling downwards)
+                if (aAllVisible && aIndex < bIndex)
+                    return -1;
+                if (bAllVisible && bIndex < aIndex)
+                    return 1;
             }
+
+            // if both are all visible - prioritize higher sections
+            if (aAllVisible && bAllVisible)
+                return aIndex.CompareTo(bIndex);
+
+            // if neither are all visible - just compare which is more visible
+            return bVisibleHeight.CompareTo(aVisibleHeight);
+        }
+
+        private float visibleHeight(T section)
+        {
+            float currentScroll = scrollContainer.Current;
+            float scrollOffset = FixedHeader?.LayoutSize.Y ?? 0;
+            float sectionPosition = scrollContainer.GetChildPosInContent(section) - scrollOffset;
+
+            // set top cutoff such that if the smallest section is above our target then the section at our target is more visible in every situation
+            // that is to say, top cutoff should be just above scroll target but not so much above that the smallest section could be 100% visible,
+            // hiding, or being more visible than, our actual target section when above it
+            float topCutoff = Math.Max(currentScroll, currentScroll + scrollContainer.DisplayableContent * scroll_target_multiplier - smallestSection.Height * 0.9f);
+            float bottomCutoff = currentScroll + scrollContainer.DisplayableContent * bottom_visibility_cutoff_multiplier;
+
+            float top = Math.Max(sectionPosition, topCutoff);
+            float bottom = Math.Min(sectionPosition + section.Height, bottomCutoff);
+            return bottom - top;
         }
```

Comment:
if you do continue work on this, remove this complete code. this should be possible with <10 loc. i cannot understand any of this.